src/core/bitmap.h

```diff
@@ -0,0 +1,62 @@
+/* basic bitmap implementation */
+typedef MVMuint64 MVMBitmap;
+
+/* Efficient find-first-set; on x86, using `bsf` primitive operation; something
+ * else on other architectures. */
+#ifdef __GNUC__
+/* also works for clang and friends */
+#define MVM_FFS(x) __builtin_ffsll(x)
+#elif defined(_MSC_VER)
+MVM_STATIC_INLINE MVMuint32 MVM_FFS(MVMBitmap x) {
+    MVMuint32 i = 0;
+    if (_BitScanForward64(&i, x) == 0)
+        return 0;
+    return i + 1;
+}
+#else
+/* fallback, note that i=0 if no bits are set */
+MVM_STATIC_INLINE MVMuint32 MVM_FFS(MVMBitmap x) {
+    MVMuint32 i = 0;
+    while (x) {
+        if (x & (1 << i++))
+            break;
+    }
+    return i;
+}
+#endif
+
+
+/* NB - make this a separate 'library', use it for register bitmap */
+/* Witness the elegance of the bitmap for our purposes. */
+MVM_STATIC_INLINE void MVM_bitmap_set(MVMBitmap *bits, MVMint32 idx) {
+    bits[idx >> 6] |= (UINT64_C(1) << (idx & 0x3f));
```

Comment:
Why 6, out of curiosity? Are we keeping the upper bits of the byte for something?

Reply:
The idx indicates the bit we want, the first 6 bits represetn 0-63, so as we're dereferencing an array of 64 bit integers, we need to shift down by 6 bits to get the correct array index